extmod/network_ninaw10.c

```diff
@@ -85,25 +134,42 @@ STATIC mp_obj_t network_ninaw10_active(size_t n_args, const mp_obj_t *args) {
                     MP_ERROR_TEXT("Failed to initialize Nina-W10 module, error: %d\n"), error);
             }
             // check firmware version
-            uint8_t fw_ver[NINA_FW_VER_LEN];
-            if (nina_fw_version(fw_ver) != 0) {
+            uint8_t semver[NINA_FW_VER_LEN];
+            if (nina_fw_version(semver) != 0) {
                 nina_deinit();
                 mp_raise_msg_varg(&mp_type_OSError,
                     MP_ERROR_TEXT("Failed to read firmware version, error: %d\n"), error);
             }
-            // Check fw version matches the driver.
-            if ((fw_ver[NINA_FW_VER_MAJOR_OFFS] - 48) != NINA_FW_VER_MAJOR ||
-                (fw_ver[NINA_FW_VER_MINOR_OFFS] - 48) != NINA_FW_VER_MINOR ||
-                (fw_ver[NINA_FW_VER_PATCH_OFFS] - 48) != NINA_FW_VER_PATCH) {
-                mp_printf(&mp_plat_print,
-                    "Warning: firmware version mismatch, expected %d.%d.%d found: %d.%d.%d\n",
-                    NINA_FW_VER_MAJOR, NINA_FW_VER_MINOR, NINA_FW_VER_PATCH,
-                    fw_ver[NINA_FW_VER_MAJOR_OFFS] - 48,
-                    fw_ver[NINA_FW_VER_MINOR_OFFS] - 48,
-                    fw_ver[NINA_FW_VER_PATCH_OFFS] - 48);
+            // Check the minimum supported firmware version.
+            uint32_t fwmin = (NINA_FW_VER_MIN_MAJOR * 100) +
+                (NINA_FW_VER_MIN_MINOR * 10) +
+                (NINA_FW_VER_MIN_PATCH * 1);
+
+            uint32_t fwver = (semver[NINA_FW_VER_MAJOR_OFFS] - 48) * 100 +
+                (semver[NINA_FW_VER_MINOR_OFFS] - 48) * 10 +
+                (semver[NINA_FW_VER_PATCH_OFFS] - 48) * 1;
+
+            if (fwver < fwmin) {
+                mp_raise_msg_varg(&mp_type_OSError,
+                    MP_ERROR_TEXT("Firmware version mismatch. Minimum supported firmware is v%d.%d.%d found v%d.%d.%d\n"),
+                    NINA_FW_VER_MIN_MAJOR, NINA_FW_VER_MIN_MINOR, NINA_FW_VER_MIN_PATCH, semver[NINA_FW_VER_MAJOR_OFFS] - 48,
+                    semver[NINA_FW_VER_MINOR_OFFS] - 48, semver[NINA_FW_VER_PATCH_OFFS] - 48);
             }
+            mp_sched_lock();
+            MP_STATE_PORT(mp_wifi_sockpoll_list) = mp_obj_new_list(0, NULL);
+            // Start sockets poll timer
+            mp_obj_t timer_args[] = {
+                MP_OBJ_NEW_QSTR(MP_QSTR_freq), MP_OBJ_NEW_SMALL_INT(10),
+                MP_OBJ_NEW_QSTR(MP_QSTR_callback), MP_OBJ_FROM_PTR(&network_ninaw10_timer_callback_obj),
+            };
+            MP_STATE_PORT(mp_wifi_timer) = machine_timer_type.make_new((mp_obj_t)&machine_timer_type, 0, 2, timer_args);
```

Comment:
Will this timer keep running if the ninaw10 is deactivated?  What if it's activated again, will it create a new timer?  Maybe it just needs a single timer that is only created once?

Reply:
I can't figure out how to deinit the timer, there's no `stop` and there's no softtimer in `rp2`. The best I could come up with is to set it to `MP_OBJ_NULL` then it will be collected and the deinitialised in the finaliser. Any better suggestions ? Can I call the finaliser directly instead ?

Reply:
> Added a check for NULL, so if the interface is reactivated it will Not recreate the timer.

Yes, that's a good solution!